File: Templates/Python/prime.py
from typing import List

class Solution:
    def __init__(self):
        """
        Initialize constants and base witnesses used in the Miller-Rabin primality tests.
        """
        self.bitmask_primes_2_to_31 = 0x208A28Ac
        self.wheel = [6, 4, 2, 4, 2, 4, 6, 2]
        self.R1 = [377687, -1]
        self.R2 = [31, 73, -1]
        self.R3 = [2, 7, 61, -1]
        self.R4 = [2, 13, 23, 1662803, -1]
        self.R5 = [2, 325, 9375, 28178, 450775, 9780504, 1795265022, -1]

    def powmod(self, base: int, exponent: int, modulo: int) -> int:
        """
        Computes (base^exponent) mod modulo using binary exponentiation.

        Args:
            base (int): The base of the exponentiation.
            exponent (int): The exponent of the exponentiation.
            modulo (int): The modulo for the operation.

        Returns:
            int: The result of (base^exponent) mod modulo.
        """
        result = 1
        while exponent > 0:
            if exponent & 1:
                result = (result * base) % modulo
            base = (base * base) % modulo
            exponent >>= 1
        return result

    def mulmod(self, a: int, b: int, modulo: int) -> int:
        """
        Performs (a * b) mod modulo safely to avoid overflow.

        Args:
            a (int): The first operand.
            b (int): The second operand.
            modulo (int): The modulo for the operation.

        Returns:
            int: The result of (a * b) mod modulo.
        """
        return (a * b) % modulo

    def gcd(self, a: int, b: int) -> int:
        """
        Calculates the Greatest Common Divisor of a and b using Euclid's algorithm.

        Args:
            a (int): The first number.
            b (int): The second number.

        Returns:
            int: The GCD of a and b.
        """
        while a != 0:
            a, b = b % a, a
        return b

    def co_prime(self, a: int, b: int) -> bool:
        """
        Checks if two numbers a and b are co-prime, which means their GCD is 1.

        Args:
            a (int): The first number.
            b (int): The second number.

        Returns:
            bool: True if a and b are co-prime, False otherwise.
        """
        return self.gcd(a, b) == 1

    def lcm(self, a: int, b: int) -> int:
        """
        Computes the Least Common Multiple of a and b.

        Args:
            a (int): The first number.
            b (int): The second number.

        Returns:
            int: The LCM of a and b.
        """
        return a * (b // self.gcd(a, b))

    def wheel_factorization(self, x: int) -> bool:
        """
        Checks if a number is prime using wheel factorization method.

        Args:
            x (int): The number to check for primality.

        Returns:
            bool: True if x is prime, False otherwise.
        """
        if x in [2, 3, 5]:
            return True
        if x % 2 == 0 or x % 3 == 0 or x % 5 == 0:
            return False
        pos = 1
        i = 7
        while i * i <= x:
            if x % i == 0:
                return False
            i += self.wheel[pos]
            pos = (pos + 1) & 7
        return x > 1

    def miller_rabin(self, n: int) -> bool:
        """
        Performs the Miller-Rabin probabilistic primality test.

        Args:
            n (int): The number to check for primality.

        Returns:
            bool: True if n is probably prime, False if n is composite.
        """
        if n < 2:
            return False
        if n != 2 and n % 2 == 0:
            return False
        if n < 31:
            return (self.bitmask_primes_2_to_31 & (1 << n)) != 0

        witnesses = self.R5 if n < 1 << 64 else [2]  # For numbers > 2^64, use 2 as a simple witness

        d, r = n - 1, 0
        while d % 2 == 0:
            r += 1
            d //= 2

        for a in witnesses:
            if a == -1:
                break
            x = pow(a, d, n)
            if x == 1 or x == n - 1:
                continue
            for _ in range(r - 1):
                x = pow(x, 2, n)
                if x == n - 1:
                    break
            else:
                return False
        return True


s = Solution()

for i in range(2, 10 ** 9 + 15):
    if s.wheel_factorization(i) != s.miller_rabin(i):
        print(f'{i} {s.wheel_factorization(i)} {s.miller_rabin(i)}')

File: Templates/C++/prime.cpp
#include <functional>
#include <iostream>
#include <regex>
#include <string>
using namespace std;

class Solution {
public:
    /* helper functions */
    int powmod(int base, int exponent, int modulo) {
        int result = 1;
        while(exponent > 0) {
            if(exponent & 1) {
                result = (result * (unsigned long long)base) % modulo;
            }
            base = (base * (unsigned long long)base) % modulo;
            exponent >>= 1;
        }
        return result;
    }

    int mulmod(int a, int b, int modulo) {
        return ((unsigned long long)a * b) % modulo;
    }

    /* template gcd */
    template <typename T>
    T gcd(T a, T b) {
        while(a != 0) {
            T c = a;
            a = b % a;
            b = c;
        }
        return b;
    }

    /* template co-prime */
    template <typename T>
    bool coPrime(T a, T b) {
        if(((a|b) & 1) == 0) {
            return false;
        }
        return gcd(a, b) == 1;
    }

    /* template lcm */
    template <typename T>
    T lcm(T a, T b) {
        return a * (b / gcd(a, b));
    }

    /* prime sieve from 2-size */
    vector<bool> primeSieve(int size) {
        const int half = (size >> 1)  + 1;
        vector<bool> sieve(half, true);
        sieve[0] = false;

        for(int i = 1; 2*i*i < half; i++) {
            if(sieve[i]) {
                int current = 3*i+1;
                while(current < half) {
                    sieve[current] = false;
                    current += 2*i+1;
                }
            }
        }

        return sieve;
    }

    /* slower but accurate */
    bool wheelFactorization(long long x) {
        if(x == 2 || x == 3 || x == 5) {
            return true;
        }
        if(x % 2 == 0 || x % 3 == 0 || x % 5 == 0) {
            return false;
        }

        const int d[] = {6, 4, 2, 4, 2, 4, 6, 2};
        int pos = 1;
        long long i = 7;

        while(i * i <= x) {
            if(x % i == 0) {
                return false;
            }
            i += d[pos];
            pos = (pos + 1) & 7;
        }
        return x > 1;
    }

    /* faster but probabilitiec primality */
    bool millerRabin(long long x) {
        if(x < 31) {
            const int bitmaskPrimes2to31 = 0x208A28Ac;
            return (bitmaskPrimes2to31 & (1 << x)) != 0;
        }

        if(x == 1 || x == 2) {
            return true;
        }

        if(x % 2 == 0 || x % 3 == 0 || x % 5 == 0 || x % 7 == 0 || x % 11 == 0 || x % 13 == 0 || x % 17 == 0) {
            return false;
        }

        if(x < 17*19) {
            return true;
        }

        const int *r, R1[] = {377687, -1}, R2[] = {31, 73, -1}, R3[] = {2, 7, 61, -1}, R4[] = {2, 13, 23, 1662803, -1}, R5[] = {2, 325, 9375, 28178, 450775, 9780504, 1795265022, -1};

        if(x < 5329) r = R1;
        else if(x < 9080191) r = R2;
        else if(x < 4759123141LL) r = R3;
        else if(x < 1122004669633LL) r = R4;
        else r = R5;

        long long d = x-1;
        d >>= 1;

        int shift = 0;
        while((d & 1) == 0) {
            shift++;
            d >>= 1;
        }

        do {
            int t = powmod(*r++, d, x);
            if(t == 1 || t == x - 1) continue;
            bool maybePrime = false;
            for(int i = 0; i < shift; i++) {
                t = mulmod(t, t, x);
                if(t == 1) return false;
                if(t == x - 1) {
                    maybePrime = true;
                    break;
                }
            }
            if(!maybePrime) return false;
        } while(*r != -1);
        return true;
    }
};

int main() {
    Solution s;
    for(int i = 2; i <= 10000005; i++) {
        if(s.millerRabin(i) != s.wheelFactorization(i)) {
            cout << i << " " << s.millerRabin(i) << " " << s.wheelFactorization(i) << "\n";
        }
    }
}
